src/beamlime/constructors/providers.py

```diff
@@ -233,10 +233,23 @@ def __init__(
         from functools import lru_cache
 
         super().__init__(_constructor, *args, **kwargs)
+        self._check_args_hashability()
+
         self.cached_result: Product
         self.cached_args = CachedArguments()
         self.cache_indicator = lru_cache(maxsize=2)(self.cached_args)
 
+    def _check_args_hashability(self) -> None:
+        if unhashable_args := {
+            arg_name: arg_spec
+            for arg_name, arg_spec in self.arg_dep_specs.items()
+            if arg_spec.runtime_type.__hash__ is None
+        }:
+            raise TypeError(
```

Comment:
In what kind of situation do you get an unhashable type? Is it relatively common? I.e. do we need to come up with an alternative handling instead of raising an error?

I guess coming up with an alternative may require fundamental changes to the rest of the code base?
Or maybe it is simply because we are using `lru_cache` and we could implement our own caching instead?

(sorry lots of questions...)

Reply:
>  do we need to come up with an alternative handling instead of raising an error?
Probably. I just came across this case where;
One of the arguments was not hashable, but it is from a cached provider, so it is guaranteed that it's the same input.
In this case, it shouldn't need to be checked if it is hashable or not.

> Or maybe it is simply because we are using lru_cache and we could implement our own caching instead?
It is internallly using `lru_cache` and `lru_cache` doesn't allow unhashable arguments. That was why it is checking if it is hashable or not.

> (sorry lots of questions...)
Not at all : D...! Thank you for reviewing